src/mono/wasm/runtime/marshal-to-js.ts

```diff
@@ -216,11 +217,58 @@ function _marshal_delegate_to_js(arg: JSMarshalerArgument, _?: MarshalerType, re
 }
 
 export class TaskHolder {
-    constructor(public resolve_or_reject: (type: MarshalerType, argInner: JSMarshalerArgument) => void) {
+    constructor(public promise: Promise<any>, public resolve_or_reject: (type: MarshalerType, js_handle: JSHandle, argInner: JSMarshalerArgument) => void) {
     }
 }
 
 export function marshal_task_to_js(arg: JSMarshalerArgument, _?: MarshalerType, res_converter?: MarshalerToJs): Promise<any> | null {
+    // this path is used only when Task is passed as argument to JSImport and virtual JSHandle would be used
+
+    // if there is synchronous result, return it
+    const promise = sync_marshal_task_to_js(arg, res_converter);
+    if (promise !== false) {
+        return promise;
+    }
+
+    const jsv_handle = get_arg_js_handle(arg);
```

Comment:
i don't like that `get_arg_js_handle` returns a jsv_handle, yet `mono_wasm_get_js_handle` returns a js_handle. it's magical and confusing to the reader that there's no type system or naming convention hints here to make it evident whether a given piece of code is correct. i don't expect you to fix that in this PR though

Reply:
Hmm, re-reading your comment, you are right. `JSVHandle` is special type of `JSHandle` on logical level. (recognized by negative values). We currently only use it pro promises/tasks.

Reply:
The usage of `TaskPreCreated` allowed me to make usage bit clearer. But still they are (and should be) interchangeable.